Python/flowgraph.c

```diff
@@ -1544,10 +1717,73 @@ optimize_if_const_binop(basicblock *bb, int i, PyObject *consts, PyObject *const
         PyErr_Clear();
         return SUCCESS;
     }
-    int newopcode, newoparg;
-    RETURN_IF_ERROR(newop_from_folded(newconst, consts, const_cache, &newopcode, &newoparg));
-    nop_out(bb, i-1, 2);
-    INSTR_SET_OP1(binop, newopcode, newoparg);
+    RETURN_IF_ERROR(instr_make_load_const(binop, newconst, consts, const_cache));
+    nop_out(bb, i-1, BINOP_OPERAND_COUNT);
+    return SUCCESS;
+}
+
+static PyObject*
+const_folding_unary_not(PyObject *v)
+{
+    int r = PyObject_IsTrue(v);
+    if (r < 0)
+        return NULL;
+    return PyBool_FromLong(!r);
+}
+
+static PyObject *
+eval_const_unaryop(PyObject *operand, int op)
+{
+    assert(operand != NULL);
+    PyObject *result;
+    switch (op) {
+        case UNARY_NEGATIVE:
+            result = PyNumber_Negative(operand);
+            break;
+        case UNARY_INVERT:
+            result = PyNumber_Invert(operand);
+            break;
+        case UNARY_NOT:
+            result = const_folding_unary_not(operand);
+            break;
+        default:
+            Py_UNREACHABLE();
+    }
+    return result;
+}
+
+static int
+optimize_if_const_unaryop(basicblock *bb, int i,
+                          PyObject *consts, PyObject *const_cache)
+{
+    #define UNARYOP_OPERAND_COUNT 1
+    cfg_instr *instr = &bb->b_instr[i];
+    assert(
+        instr->i_opcode == UNARY_NEGATIVE
+        || instr->i_opcode == UNARY_INVERT
+        || instr->i_opcode == UNARY_NOT
+        || (instr->i_opcode == CALL_INTRINSIC_1 && instr->i_oparg == INTRINSIC_UNARY_POSITIVE)
+    );
+    PyObject *seq;
+    RETURN_IF_ERROR(get_constant_sequence(bb, i-1, UNARYOP_OPERAND_COUNT, consts, &seq));
+    RETURN_SUCCESS_IF_NULL(seq);
+    assert(PyTuple_Size(seq) == UNARYOP_OPERAND_COUNT);
+    PyObject *operand = PyTuple_GET_ITEM(seq, 0);
+    bool unarypos = instr->i_opcode == CALL_INTRINSIC_1;
+    PyObject *newconst = unarypos ? PyNumber_Positive(operand) : eval_const_unaryop(operand, instr->i_opcode);
```

Comment:
I'd add an oparg parameter to `eval_const_unaryop` so that `CALL_INTRINSIC_1` doesn't need to be special-cased here. The assertion in lines 1761-6 can probably move there too.